utils/fork-tree/src/lib.rs

```diff
@@ -696,43 +697,56 @@ mod node_implementation {
 
 		pub fn import<F, E: std::error::Error>(
 			&mut self,
-			mut hash: H,
-			mut number: N,
-			mut data: V,
+			hash: H,
+			number: N,
+			data: V,
 			is_descendent_of: &F,
 		) -> Result<Option<(H, N, V)>, Error<E>>
 			where E: fmt::Debug,
 				  F: Fn(&H, &H) -> Result<bool, E>,
 		{
-			if self.hash == hash {
-				return Err(Error::Duplicate);
-			};
+			let mut stack: Vec<(&Self, bool)> = vec![(&self, true)];
 
-			if number <= self.number { return Ok(Some((hash, number, data))); }
+			while let Some((node, first_time)) = stack.pop() {
+				if first_time {
+					if node.hash == hash {
+						return Err(Error::Duplicate);
+					}
 
-			for node in self.children.iter_mut() {
-				match node.import(hash, number, data, is_descendent_of)? {
-					Some((h, n, d)) => {
-						hash = h;
-						number = n;
-						data = d;
-					},
-					None => return Ok(None),
+					if number <= node.number {
+						continue;
+					}
+
+					if !node.children.is_empty() {
+						// We want to check all children depth-first.
+
+						stack.push((node, false));
+
+						for child in node.children.iter().rev() {
+							stack.push((child, true));
+						}
+
+						continue;
+					}
 				}
-			}
 
-			if is_descendent_of(&self.hash, &hash)? {
-				self.children.push(Node {
-					data,
-					hash: hash,
-					number: number,
-					children: Vec::new(),
-				});
-
-				Ok(None)
-			} else {
-				Ok(Some((hash, number, data)))
+				if is_descendent_of(&node.hash, &hash)? {
+					let node = &mut (node as *const Self as *mut Self);
+
+					unsafe {
+						(**node).children.push(Node {
+							data,
+							hash: hash,
+							number: number,
+							children: Vec::new(),
+						});
+					}
```

Comment:
We really shouldn't b doing this unsafeness here, but I'm not sure what the solution is. Do we just wrap everything in locks?

Reply:
I really don't know how to avoid unsafe here, I only have one way of doing it, it is to use some struct with pointer on it and recursively call so the double mut pointer needed is handled by function call, but that totally defeat the point here (I did stumble upon the same thing when writing tree code not writing recursive code, and the alternative crates doing the same thing safely does it recursing). So I am not very optimistic on avoiding unsafe code here.
About this unsafe usage, there is an alternative, with a little rewrite you can limit your unsafety to lifetime: by storing `*mut Self` in the stack, you do not have `*const` casted to `*mut`.
But you still need some unsafe cal to 'as_mut' to cast '* mut' to '&mut', not sure if it is really better (I propose it because you do not have to think about possible *const optim, but with a single update it is probably fine).